### PR TITLE
Use a class with implicit overflow checks for Address

### PR DESCRIPTION
**Pull request by dschuff**

### CODE REVIEW

File: src/s2wasm.h [hunk 1]
```diff
@@ -1071,7 +1072,7 @@ class S2WasmBuilder {
           value = getInt();
           if (value != 0) zero = false;
         }
-        for (Address i = 0, e = size; i < e; i++) {
+        for (Address i = 0, e = size; i < e; ++i) {
```
Comment (kripken) [inline src/s2wasm.h]: why?

File: src/wasm.h [hunk 2]
```diff
@@ -1386,5 +1400,12 @@ class Module {
 };
 
 } // namespace wasm
+namespace std {
```
Comment (kripken) [inline src/wasm.h]: empty line before new namespace please

File: src/wasm.h [hunk 3]
```diff
@@ -92,7 +92,21 @@ struct Name : public cashew::IString {
...
 // An address in linear memory. For now only wasm32
-typedef uint32_t Address;
+struct Address {
+  typedef uint32_t address_t;
+  address_t addr;
+  Address() : addr(0) {}
+  Address(uint64_t a) : addr(static_cast<address_t>(a)) {
+    assert(a <= std::numeric_limits<address_t>::max());
+  }
+  Address& operator =(uint64_t a) {
```
Comment (kripken) [inline src/wasm.h]: we don't have spaces after `operator` in other places. but this does look nice. just making sure it's not unintentional.

**Comment by kripken** [comment]
lgtm

File: src/wasm.h (hunk 3 above)
Comment (dschuff) [inline src/wasm.h]: I like having spaces better in general, but consistency is better here I think. I'll change it.

**Comment by dschuff** [comment]
Retroactively, this is related to #278

**Pull request merged**